src/backend/distributed/master/master_metadata_utility.c

```diff
@@ -57,6 +61,271 @@ static void RecordDistributedRelationDependencies(Oid distributedRelationId,
+	List *shardIntervalList = NIL;
+	int shardIndex = 0;
+	int shardIntervalArrayLength = distTableCacheEntry->shardIntervalArrayLength;
+
+	for (shardIndex = 0; shardIndex < shardIntervalArrayLength; shardIndex++)
+	{
+		ShardPlacement *placementArray =
+			distTableCacheEntry->arrayOfPlacementArrays[shardIndex];
+		int numberOfPlacements =
+			distTableCacheEntry->arrayOfPlacementArrayLengths[shardIndex];
+		int placementIndex = 0;
+
+		for (placementIndex = 0; placementIndex < numberOfPlacements; placementIndex++)
+		{
+			ShardPlacement *placement = &placementArray[placementIndex];
+			char *shardNodeName = placement->nodeName;
+			uint32 shardNodePort = placement->nodePort;
+			uint64 shardId = placement->shardId;
+			bool metadataLock = false;
+
+			metadataLock = TryLockShardDistributionMetadata(shardId, ShareLock);
+
+			/* if the lock is not acquired warn the user */
+			if (metadataLock == false)
+			{
+				ereport(WARNING, (errcode(ERRCODE_LOCK_NOT_AVAILABLE),
+								  errmsg("lock is not acquired, size of shard %ld "
+										 "will be ignored", shardId)));
+				continue;
+			}
+
+			if (strcmp(shardNodeName, workerNodeName) == 0 &&
+				shardNodePort == workerNodePort)
+			{
+				ShardInterval *shardInterval =
+					distTableCacheEntry->sortedShardIntervalArray[shardIndex];
+				shardIntervalList = lappend(shardIntervalList, shardInterval);
+			}
+		}
+	}
+
+	return shardIntervalList;
+}
+
+
+/*
+ * GenerateSizeQueryOnMultiplePlacements generates a select size query to get
+ * size of multiple tables from the relation with distributedRelationId. Note that,
+ * different size functions supported by PG are also supported by this function changing
+ * the size query given as the last parameter to function. Format of sizeQuery is
+ * pg_*_size(%s). Examples of it can be found in the master_protocol.h
+ */
+static StringInfo
+GenerateSizeQueryOnMultiplePlacements(Oid distributedRelationId, List *shardIntervalList,
+									  char *sizeQuery)
+{
+	/* schema name will be used to form shard qualified name */
+	Oid schemaId = get_rel_namespace(distributedRelationId);
+	char *schemaName = get_namespace_name(schemaId);
+
+	StringInfo selectQuery = makeStringInfo();
+	ListCell *shardIntervalCell = NULL;
+
+	appendStringInfo(selectQuery, "SELECT ");
+
+	foreach(shardIntervalCell, shardIntervalList)
+	{
+		ShardInterval *shardInterval = (ShardInterval *) lfirst(shardIntervalCell);
+		uint64 shardId = shardInterval->shardId;
+		char *shardName = get_rel_name(distributedRelationId);
+		char *shardQualifiedName = NULL;
+		char *quotedShardName = NULL;
+		AppendShardIdToName(&shardName, shardId);
+
+		shardQualifiedName = quote_qualified_identifier(schemaName, shardName);
+		quotedShardName = quote_literal_cstr(shardQualifiedName);
+
+		appendStringInfo(selectQuery, sizeQuery, quotedShardName);
+		appendStringInfo(selectQuery, " + ");
+	}
+
+	/* there is no more size function so just add 0 instead of it */
+	appendStringInfo(selectQuery, "0;");
```

Comment:
Adding 0 here is very hacky. It would be hard for somebody who looks into worker logs to understand why we add 0. You can check the length of the list and skip adding + at the end.
An example; https://github.com/citusdata/citus/blob/0a58a7b2c1f681fae350cf5d83c1b4c5a058e8bb/src/backend/distributed/metadata/metadata_sync.c#L424